src/core/Akka/IO/TcpConnection.cs

```diff
@@ -815,73 +827,169 @@ private sealed class PendingBufferWrite : PendingWrite
                 IActorRef self,
                 IActorRef commander,
                 ByteString buffer,
-                object ack,
-                WriteCommand tail) : base(commander, ack)
+                object ack) : base(commander, ack)
             {
                 _connection = connection;
                 _sendArgs = sendArgs;
                 _self = self;
                 _buffer = buffer;
-                _tail = tail;
 
                 // start immediatelly as we'll need to cover the case if 
                 // after buffer write request, the remaining enumerator is empty
                 //hasData = this.remainingData.MoveNext();
             }
 
-            public override PendingWrite DoWrite(ConnectionInfo info)
+            public override void DoWrite(ConnectionInfo info)
             {
                 try
                 {
-                    var data = _buffer;
-                    while(true)
-                    {
-                        var bytesWritten = Send(data);
-
-                        if (_connection.traceLogging)
-                            _connection.Log.Debug("Wrote [{0}] bytes to channel", bytesWritten);
-                        if (bytesWritten < data.Count)
-                        {
-                            // we weren't able to write all bytes from the buffer, so we need to try again later
-                            data = data.Slice(bytesWritten);
-                        }
-                        else // finished writing
-                        {
-                            if(Ack != NoAck.Instance) Commander.Tell(Ack);
-                            Release();
-                            return _connection.CreatePendingWrite(Commander, _tail, info);
-                        }
-                    }
-
+                    _sendArgs.SetBuffer(_buffer);
+                    if (!_connection.Socket.SendAsync(_sendArgs))
+                        _self.Tell(SocketSent.Instance);
+                    
+                    Release();
```

Comment:
Is the idea behind the `Release` method here to make the `PendingBufferWrite`s re-usable in a pool, so we don't have to frequently allocate and GC new ones? CC @Horusiath

Reply:
@Aaronontheweb Looking at scala's implementation, here is what they do: they are taking `ByteString` that should be transferred, then allocating buffer from buffer pool, and make copy of the bytestring to that buffer. And after `Send` is complete, they are using `Release` method to return this buffer back to the pool.

Our `ByteString` is build on top of `ArraySegment<byte>`, which in turn can be set as `SocketAsyncEventArgs` source buffer for sending in `SendAsync`, without any allocations. I do not see any point is making copy to the buffer and releasing it that (maybe this is done due to some side-effect of scala's socket `Send` implementation?). 

So, I am suggesting to just remove this Release method, because we are not allocating anything here, and nothing to be released.

Reply:
Update: Just realized that to make buffered writes sent all at once I will need to allocate buffer to store all of them in single array - so maybe Release method will still be relevant. Will see.

Reply:
Yeah it sounds like object pooling is what they're doing there. We should probably take a look at making `ByteString` castable to something like `Span<T>` or `Memory<T>` in a future release.

Reply:
One more update: turned out that `SocketAsyncEventArgs` are using same `IList<ArraySegment<byte>>` as our `ByteString` does, so no more allocations were required to pass data into socket. Nice.

Reply:
Very nice